### ISSUE TITLE
Streamlining the API

### ISSUE THREAD

**Issue by bmcfee** [marl/jams#29]
Still in the interest of minimizing the effort for #26 , I'm looking for ~~corners to cut~~ places in which the API can be streamlined, IMO.

Since this keeps recurring, I figured we should have a dedicated thread.

* [`AnnotationArray.create_annotation()`](https://github.com/marl/jams/blob/develop/jams/pyjams.py#L611)  -- construct a new `Annotation`, append it to the end of an existing `AnnotationArray`, and return the new annotation.

This is a minor convenience method.  Since both the AnnotationArray and the new Annotation will be in the caller's scope afterwards, I don't see why it's necessary.  For instance, the following are equivalent:

```python
>>> new_ann = jams.Annotation(namespace='foo', data='bar')
>>> existing_jam.annotations.append(new_ann)
```
vs.
```python
>>> new_ann = existing_jam.annotations.create_annotation(namespace='foo', data='bar')
```
I prefer the former because it's entirely explicit about what's created and where.  In the latter case, appending to the annotation array is a side-effect of construction, which feels a little messy.

The only case in which it would make a qualitative difference is when the return value of `create_annotation` is ignored, but you can get the same effect by saying
```python
>>> existing_jam.annotations.append(jams.Annotation(namespace='foo', data='bar'))
```
which I think is also more clear, and more "pythonic" in that it uses existing API (inherited from list) rather than creating new terminology.

Any counter-arguments?

**Comment by ejhumphrey** [comment]
I think this convenience factory method matters less now, and is a vestige
of the days when there were different Annotation / Observation types. Now
that it's all the same, I agree it can be done away with.

**Comment by bmcfee** [comment]
Done.

Another one, and this may disappear after #27 , but it seems like `JObject.__eq__` can be simplified to the following:
```python
def __eq__(self, other):
    return isinstance(other, self.__class__) and (self.__dict__ == other.__dict__)
```
This would bypass all serialization, and should thus be much faster.  (It also lets python propagate the appropriate equality methods all the way through the object hierarchy, which seems like a good thing when it comes down to JamsFrames.)

**Comment by urinieto** [comment]
I would also support the former case.

**Comment by bmcfee** [comment]
Fixed as of fe0f68cc83333e2cf958437e37faf8ce5766e602 : relying on serialization for comparison caused some unpredictable test behavior.  Going directly to the object equality fixes it.

Also, JamsFrame now uses `DataFrame.equals()` for its equality operation.  I'm not sure if this is a great idea, since it breaks the default behavior of DataFrame (element-wise equality).  It might be better to push this comparison check up a level to the Annotation class, and leave JamsFrame as close to DataFrame as possible.  Any thoughts?

**Comment by bmcfee** [comment]
New legit-question time: do we need AnnotationArray?

Arguments in favor of keeping AnnotationArray:
- It's already there, and it works
- It helps contain the search and serialization logic for the annotations
- The serialization logic might be simpler this way

Arguments in favor of ditching it:
- It doesn't correspond to a definition in the JAMS schema
- All it implements now is search and serialization
- `JAMS.search()` is just a wrapper for `AnnotationArray.search()`.  Why not move the search logic up a level?
- Serialization should be automatic anyway
- Any JAMS object has exactly one AnnotationArray object now, so why have a separate class, instead of a primitive (list)?

**Comment by bmcfee** [comment]
[tagging everyone in hopes of commentary] @ejhumphrey @justinsalamon @urinieto @rabitt

**Comment by urinieto** [comment]
When working with multiple annotations it is quite nice to have easy access
and iteration to the all the annotations. I guess as long as something
similar is available (e.g., list), we could ditch that object.

**Comment by bmcfee** [comment]
> When working with multiple annotations it is quite nice to have easy access
> and iteration to the all the annotations.

Yes, all I'm asking is whether we need a custom list object anymore.  All it currently does is serialize element-wise (which I think is redundant) and provide a search interface: https://github.com/marl/jams/blob/develop/jams/pyjams.py#L988 .  Since the top-level JAMS object short-circuits to the annotation array search, I think we can simplify this.

**Comment by bmcfee** [comment]
I took a stab at this yesterday in a branch, and it was a total disaster from a serialization perspective.  I think we're probably fine keeping AnnotationArray around.

Otherwise, I'm pretty happy with the API at this point.  Are there any objections to closing out this thread?

**Comment by urinieto** [comment]
Not from my part. I'm also pretty happy with the state of JAMS, and
considering using this version on MSAF already.
